---
title: "*SignalingProfiler* 2.0 Step 1 benchmarking"
author: "Veronica Venafra"
output:
  html_document:
    theme: flat
    highlight: zenburn
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

## Introduction
This workbook refers to the manuscript *SignalingProfiler 2.0: a network-based approach to bridge multi-omic data to phenotypic hallmarks*.

## Import libraries and functions

```{r message = FALSE}
source('0.libraries.R')
source('0.functions.R')
source('0.heatmap_drawer.R') # Function to draw heatmaps for gold standard comparison
```

# Run footprint-based analyses

### Transcription Factor Enrichment Analysis

To benchmark the Transcription Factor Enrichment Analysis we run SignalingProfiler analysis with:

-   4 different databases as TF - target genes regulons source: Dorothea, Dorothea+SIGNOR, CollecTRI and CollecTRI + SIGNOR

-   with and without **Hypergeometric Correction**

-   with and without **VIPER Correction with Proteomics**

```{r eval = FALSE}
# Read omic data
tr_df <- read_tsv('./input/transcriptomics.tsv')
prot_df <- read_tsv('./input/proteomics.tsv')

# Set output directories
output_dir <- './results/3.protein_activity_inference/tfea/'

# Set all possible parameters conditions
hypergeometric_params <- c(TRUE, FALSE)
names(hypergeometric_params) <- paste0('Hyp_', c(TRUE, FALSE))
correct_VIPER <- c(TRUE, FALSE)
names(correct_VIPER) <- paste0('correct_', c(TRUE, FALSE))
tfea_dbs <- c('Doro', 'DoroSign', 'Coll', 'CollSign')

list('database' = tfea_dbs,
     'hypergeom' = hypergeometric_params,
     'VIPER_correction' = correct_VIPER) -> tfea_params

expand.grid(tfea_params) -> tfea_conditions

for(tfea_db in tfea_dbs){
  for(hypergeometric_param in hypergeometric_params){
    for(correct_VIPER_i in correct_VIPER){


      output <- run_footprint_based_analysis(omic_data = tr_df,
                                   analysis = 'tfea',
                                   organism = 'human',
                                   reg_minsize = 10,
                                   exp_sign = FALSE,
                                   integrated_regulons = FALSE,
                                   collectri = FALSE,
                                   benchmark = tfea_db,
                                   hypergeom_corr = hypergeometric_param,
                                   correct_proteomics = correct_VIPER_i,
                                   prot_df = prot_df,
                                   GO_annotation = TRUE)


      write_tsv(output, paste0(output_dir, 'tfea_', tfea_db, '_',
                               names(correct_VIPER[correct_VIPER == correct_VIPER_i]), '_',
                               names(hypergeometric_params[hypergeometric_params == hypergeometric_param]),
                               '.tsv'))
    }
  }
}

# ===================================================================== #
# Process tfea output
# ===================================================================== #

hypergeometric_params <- c(T, F)
names(hypergeometric_params) <- paste0('Hyp_', c(T, F))
correct_VIPER <- c(T, F)
names(correct_VIPER) <- paste0('correct_', c(T, F))
tfea_dbs <- c('Doro', 'DoroSign', 'Coll', 'CollSign')

for(tfea_db in tfea_dbs){
  for(hypergeometric_param in hypergeometric_params){
    for(correct_VIPER_i in correct_VIPER){

      # Read inferred data
      output_tfea <- read_tsv(paste0(output_dir, 'tfea_', tfea_db, '_',
                                     names(correct_VIPER[correct_VIPER == correct_VIPER_i]), '_',
                                     names(hypergeometric_params[hypergeometric_params == hypergeometric_param]),
                                     '.tsv'))

      if(hypergeometric_param == TRUE){

        output_tfea <- output_tfea %>% select(UNIPROT, gene_name, NES = weightedNES, pvalues, mf)

      }else{
        output_tfea <- output_tfea %>% relocate(UNIPROT, gene_name, NES, pvalues, mf)
      }

      # Rename columns according to condition of analysis
      colnames(output_tfea) <- c('UNIPROT', 'gene_name',
                                 paste0(c('NES', 'pval'),
                                        paste0('_', tfea_db, '_weight=',hypergeometric_param, '_prot=', correct_VIPER_i)),
                                 'mf')

      if(i == 1){
        protein_total_df <- output_tfea

      }else{
        protein_total_df <- full_join(protein_total_df, output_tfea, by = c('UNIPROT', 'gene_name', 'mf'))
      }

      i = i + 1

    }
  }
}

write_tsv(protein_total_df, paste0(output_dir, 'tfea_protein_total.tsv'))
```

### Kinase Substrates Enrichment Analysis

To benchmark the Kinase Substrates Enrichment Analysis we run SignalingProfiler analysis with:

-   with and without **Phosphoproteomics normalization over proteomics** (see '1.Data_preparation_for_SP.Rmd')

-   2 different databases as Kinase/Phosphatase - target phosphosites regulons source: Omnipath and Omnipath+SIGNOR

-   with and without **Hypergeometric Correction**

-   with and without **VIPER Correction with Proteomics**

```{r eval = FALSE}
# Read proteomic data  
prot_df <- read_tsv('./input/proteomics.tsv')

# Set output directories
output_dir <- './results/3.protein_activity_inference/ksea/'

# Set parameters
proteomic_corrections <- c(T,F)
integrated_regulons <- c(T, F)
hypergeometric_params <- c(TRUE, FALSE)
names(hypergeometric_params) <- paste0('Hyp_', c(TRUE, FALSE))
correct_VIPER <- c(TRUE, FALSE)
names(correct_VIPER) <- paste0('correct_', c(TRUE, FALSE))

list('database' = c('Omnipath', 'Omnipath+KinomeAtlas'),
     'hypergeom' = hypergeometric_params,
     'VIPER_correction' = correct_VIPER,
     'phospho_correction' = c(T,F)) -> ksea_params

expand.grid(ksea_params) -> ksea_conditions

for(proteomic_correction in proteomic_corrections){

  # Choose normalized or not normalized phosphoproteomic data
  if(proteomic_correction){
    phospho_df <- read_tsv('./input/corrected_phosphoproteomics.tsv')
  }else{
    phospho_df <- read_tsv('./input/phosphoproteomics.tsv')
  }

  for(integrated_regulon in integrated_regulons){
    for(hypergeometric_param in hypergeometric_params){
      for(correct_VIPER_i in correct_VIPER){

        output <- run_footprint_based_analysis(omic_data = phospho_df,
                                               analysis = 'ksea',
                                               organism = 'human',
                                               reg_minsize = 5,
                                               exp_sign = FALSE,
                                               integrated_regulons = integrated_regulon,
                                               hypergeom_corr = hypergeometric_param,
                                               correct_proteomics = correct_VIPER_i,
                                               prot_df = prot_df,
                                               GO_annotation = TRUE)

        write_tsv(output, paste0(output_dir, 'ksea_',
                                 ifelse(proteomic_correction, 'corr_phospho_', 'no_corr_phospho_'),
                                 ifelse(integrated_regulon, 'int_', 'no_int_'),
                                 names(correct_VIPER[correct_VIPER == correct_VIPER_i]), '_',
                                 names(hypergeometric_params[hypergeometric_params == hypergeometric_param]),
                                 '.tsv'))

      }
    }
  }
}

# ===================================================================== #
# Process ksea output
# ===================================================================== #

proteomic_corrections <- c(T, F)
integrated_regulons <- c(T, F)
hypergeometric_params <- c(TRUE, FALSE)
names(hypergeometric_params) <- paste0('Hyp_', c(TRUE, FALSE))
correct_VIPER <- c(TRUE, FALSE)
names(correct_VIPER) <- paste0('correct_', c(TRUE, FALSE))

i = 1
for(proteomic_correction in proteomic_corrections){
  for(integrated_regulon in integrated_regulons){
    for(hypergeometric_param in hypergeometric_params){
      for(correct_VIPER_i in correct_VIPER){

        flag_prot <- ifelse(proteomic_correction, 'corr_phospho_', 'no_corr_phospho_')
        flag_yaffe <- ifelse(integrated_regulon, 'int_', 'no_int_')

        output_ksea <- read_tsv(paste0(output_dir, 'ksea_',
                                       flag_prot,
                                       flag_yaffe,
                                       names(correct_VIPER[correct_VIPER == correct_VIPER_i]), '_',
                                       names(hypergeometric_params[hypergeometric_params == hypergeometric_param]),
                                       '.tsv'))

        if(hypergeometric_param == TRUE){

          output_ksea <- output_ksea %>% select(UNIPROT, gene_name, NES = weightedNES, pvalues, mf)

        }else{
          output_ksea <- output_ksea %>% relocate(UNIPROT, gene_name, NES, pvalues, mf)
        }

        colnames(output_ksea) <- c('UNIPROT', 'gene_name',
                                   paste0(c('NES', 'pval'),
                                          paste0('_', flag_prot,  flag_yaffe, 'weight=',hypergeometric_param, '_prot=', correct_VIPER_i)),
                                   'mf')

        if(i == 1){
          protein_total_df <- output_ksea

        }else{
          protein_total_df <- full_join(protein_total_df, output_ksea, by = c('UNIPROT', 'gene_name', 'mf'))
        }

        i = i + 1
      }
    }
  }
}

write_tsv(protein_total_df, paste0(output_dir, 'ksea_protein_total.tsv'))
```

# Run PhosphoScore analysis

To benchmark the PhosphoScore analysis we run SignalingProfiler analysis with:

-   with and without **Phosphoproteomics normalization over proteomics** (see '1.Data_preparation_for_SP.Rmd')

-   2 different types of regulatory sites **activatory set**: the first set is composed of phosphosites regulating the sole activity, the second set includes also the ones regulating stability;

```{r eval = FALSE}

# Read proteomic data  
prot_df <- read_tsv('./input/proteomics.tsv')

# Set output directories
output_dir <- './results/3.protein_activity_inference/phosphoscore/'

# Set parameters
proteomic_corrections <- c(T, F)
activatory_set <- c(T, F)

for(proteomic_correction in proteomic_corrections){

  # Choose normalized or not normalized phosphoproteomic data
  if(proteomic_correction){
    phospho_df <- read_tsv('./input/corrected_phosphoproteomics.tsv')
    #phospho_df <- read_tsv('./input/regression_phosphoproteomics.tsv')
  }else{
    phospho_df <- read_tsv('./input/phosphoproteomics.tsv')
  }

  for(activatory in activatory_set){

    phosphoscore_df <- phosphoscore_computation(phosphoproteomic_data = phospho_df,
                                                organism = 'human',
                                                activatory = activatory ,
                                                GO_annotation= TRUE)

    write_tsv(phosphoscore_df, paste0(output_dir, 'phosphoscore_',
                             ifelse(proteomic_correction, 'corr_phospho_', 'no_corr_phospho_'),
                             ifelse(activatory, 'all', 'act'),
                             '.tsv'))

  }
}

# ===================================================================== #
# Process phosphoscore output
# ===================================================================== #

i = 1
for(proteomic_correction in proteomic_corrections){

  if(proteomic_correction){
    phospho_df <- read_tsv('./input/corrected_phosphoproteomics.tsv')
    #phospho_df <- read_tsv('./input/regression_phosphoproteomics.tsv')
  }else{
    phospho_df <- read_tsv('./input/phosphoproteomics.tsv')
  }

  for(activatory in activatory_set){

    flag_prot <- ifelse(proteomic_correction, 'corr_phospho_', 'no_corr_phospho_')
    flag_activatory <- ifelse(activatory, 'act', 'all')

    output_phospho <- read_tsv(paste0(output_dir, 'phosphoscore_',
                                      flag_prot,
                                      flag_activatory,
                                      '.tsv'))

    output_phospho <- output_phospho %>% select(UNIPROT, gene_name, phosphoscore, mf)

    colnames(output_phospho) <- c('UNIPROT', 'gene_name',
                               paste0('phosphoscore',
                                      paste0('_', flag_prot,  flag_activatory)),
                               'mf')

    if(i == 1){
      protein_total_df <- output_phospho

    }else{
      protein_total_df <- full_join(protein_total_df, output_phospho, by = c('UNIPROT', 'gene_name'))
    }

    i = i + 1
  }
}

write_tsv(protein_total_df, paste0(output_dir, 'phosphoscore_protein_total.tsv'))
```

# Compute the final activity score in each molecular function

### Transcription Factors

```{r eval = FALSE}

# Set output directories
output_dir <- paste0('./results/3.protein_activity_inference/combination/')

# Read transcription factors inferred with all possible combination of parameters
output_dir_tfea <- './results/3.protein_activity_inference/tfea/'
protein_total_df_tfea <- read_tsv(paste0(output_dir_tfea, 'tfea_protein_total.tsv'))
protein_total_df_tfea_sub <- protein_total_df_tfea %>% select(UNIPROT, gene_name, contains('NES'))

# Read PhosphoScore analysis run with all possible combination of parameters
output_dir_phos <- './results/3.protein_activity_inference/phosphoscore/'
protein_total_ps <- read_tsv(paste0(output_dir_phos, 'phosphoscore_protein_total.tsv'))#

protein_total_ps_sub <- protein_total_ps %>%
  mutate(mf = coalesce(mf.x, mf.y, mf.x.x, mf.y.y)) %>%
  select(-mf.x, -mf.y, -mf.x.x, -mf.y.y)

i_cond = 1
for(i_tfea in c(3:ncol(protein_total_df_tfea_sub))){

  for(i_ps in c(3:6)){

    name_tf <- colnames(protein_total_df_tfea_sub)[i_tfea]

    tfea_sub <- protein_total_df_tfea_sub %>%
      select(UNIPROT, gene_name, weightedNES = i_tfea) %>%
      mutate(mf = 'tf') %>%
      filter(!is.na(weightedNES))

    name_ps <- colnames(protein_total_ps_sub)[i_ps]
    psp_sub <- protein_total_ps_sub %>%
      select(UNIPROT, gene_name, phosphoscore = i_ps, mf) %>%
      filter(!is.na(phosphoscore))

    com_df <- combine_footprint_and_phosphoscore(footprint_output = tfea_sub,
                                                 phosphoscore_df = psp_sub,
                                                 analysis = 'tfea')

    com_df_sub <- com_df %>% select(gene_name, final_score, method)
    colnames(com_df_sub) <- c('gene_name',
                              paste0(name_tf, '_', name_ps),
                              paste0('method_', i_cond))

    if(i_cond == 1){
      all_com_df <- com_df_sub
    }else{
      all_com_df <- full_join(all_com_df, com_df_sub, by = c('gene_name'))
    }

    i_cond = i_cond + 1
  }
}


write_tsv(all_com_df, paste0(output_dir, 'tfs_protein_total.tsv'))
```

### Kinases

```{r eval = FALSE}

# Set output directories 
output_dir <- paste0('./results/3.protein_activity_inference/combination/')
graph_dir <- paste0('./graphs/3.protein_activity_inference/combination/')

# Read kinases and phosphatases inferred with all possible combination of parameters
output_dir_ksea <- './results/3.protein_activity_inference/ksea/'#old/'
protein_total_df_ksea <- read_tsv(paste0(output_dir_ksea, 'ksea_protein_total.tsv'))
protein_total_df_ksea_sub <- protein_total_df_ksea %>% select(UNIPROT, gene_name, contains('NES'), mf)

# Read PhosphoScore analysis run with all possible combination of parameters
output_dir_phos <- './results/3.protein_activity_inference/phosphoscore/'#
protein_total_ps <- read_tsv(paste0(output_dir_phos, 'phosphoscore_protein_total.tsv'))#

protein_total_ps_sub <- protein_total_ps %>%
  mutate(mf = coalesce(mf.x, mf.y, mf.x.x, mf.y.y)) %>%
  select(-mf.x, -mf.y, -mf.x.x, -mf.y.y)

i_cond = 1
for(i_ksea in c(3:(ncol(protein_total_df_ksea_sub)-1))){
  #i_ksea = 3
  for(i_ps in c(3:6)){
    #i_ps = 3

    name_ks <- colnames(protein_total_df_ksea_sub)[i_ksea]

    ksea_sub <- protein_total_df_ksea_sub %>%
      select(UNIPROT, gene_name, weightedNES = i_ksea, mf) %>%
      filter(!is.na(weightedNES))

    name_ps <- colnames(protein_total_ps_sub)[i_ps]
    psp_sub <- protein_total_ps_sub %>%
      select(UNIPROT, gene_name, phosphoscore = i_ps, mf) %>%
      filter(!is.na(phosphoscore))

    # Skip the condition of normalization in ksea and normalization in phosphoscore combination
    if(grepl('NES_no_corr_phospho', name_ks) & grepl('phosphoscore_corr', name_ps) | grepl('NES_corr_phospho', name_ks) & grepl('phosphoscore_no_corr', name_ps)){
      next
    }

    com_df <- combine_footprint_and_phosphoscore(footprint_output = ksea_sub,
                                                 phosphoscore_df = psp_sub,
                                                 analysis = 'ksea')

    com_df_sub <- com_df %>% select(gene_name, final_score, method)
    colnames(com_df_sub) <- c('gene_name',
                              paste0(name_ks, '_', name_ps),
                              paste0('method_', i_cond))

    if(i_cond == 1){
      all_com_df <- com_df_sub
    }else{
      all_com_df <- full_join(all_com_df, com_df_sub, by = c('gene_name'))
    }

    i_cond = i_cond + 1
  }
}

write_tsv(all_com_df, paste0(output_dir, 'kinphos_protein_total.tsv'))
```

### Other signaling proteins

```{r eval = FALSE}
# Set output directories 
output_dir <- './results/3.protein_activity_inference/combination/'
output_dir_phos <- './results/3.protein_activity_inference/phosphoscore/'

# PhosphoScore analysis run with all possible combination of parameters
protein_total_ps <- read_tsv(paste0(output_dir_phos, 'phosphoscore_protein_total.tsv'))#

protein_total_ps_sub <- protein_total_ps %>%
  mutate(mf = coalesce(mf.x, mf.y, mf.x.x, mf.y.y)) %>%
  select(-mf.x, -mf.y, -mf.x.x, -mf.y.y)

protein_total_ps_sub_other <- protein_total_ps_sub %>% filter(mf == 'other')
protein_total_ps_sub_other$mf <- NULL
protein_total_ps_sub_other$UNIPROT <- NULL

NES_df <- protein_total_ps_sub_other

write_tsv(NES_df, paste0(output_dir, 'other_protein_total.tsv'))
```

# Visualize benchmark results for Step 1

This code chunk compares each combination of parameters of the three techniques of SignalingProfiler with the gold standard proteins.

```{r}
downstream <- read_excel('./input/Gold standard.xlsx', sheet = 1, skip = 1)
regulators <- downstream %>%
  mutate(Regulation = ifelse(grepl('MTOR*', Regulator), Regulation*-100, Regulation*100)) %>%
  filter(is.na(Remove)) %>%
  distinct()
head(regulators)
```

Set colors for the heatmaps

```{r}
paletteLength <-1000
RdBu <- RColorBrewer::brewer.pal(n = 11, 'RdBu')
myColor <- colorRampPalette(c(RdBu[10], "white", RdBu[2]))(paletteLength)
```

### Transcription Factors

```{r fig.height = 4, fig.width = 8, fig.align='center'}

# Set output directories 
output_dir <- paste0('./results/3.protein_activity_inference/combination/')
output_graph <- paste0('./graphs/3.protein_activity_inference/combination/')

# Read the transcription factors inferred with TFEA, PhosphoScore or both
# and transform the tibble in a matrix without NAs
all_com_df <- read_tsv(paste0(output_dir, 'tfs_protein_total.tsv'))
NES_all <- all_com_df %>% dplyr::select(gene_name, contains('NES'))
NES_mat <- NES_all %>% column_to_rownames('gene_name')
NES_mat[is.na(NES_mat)] <- 0

# Define annotation tables for the different technical conditions
annotation_df <- tibble(conditions = colnames(NES_all)[-1], sample = paste0('Sample', 1:(ncol(NES_all)-1)))

annotation_df_plot <- annotation_df %>%
  mutate('Database' = ifelse(grepl('DoroSign', conditions), 'Dorothea+SIGNOR',
                             ifelse(grepl('_Doro_', conditions ), 'Dorothea',
                                    ifelse(grepl('Coll_', conditions), 'CollecTRI', 'CollecTRI+SIGNOR'))),
         'Hypergeometric test' = ifelse(grepl('weight=TRUE', conditions), 'TRUE', 'FALSE'),
         'VIPER correction with proteomics' = ifelse(grepl('prot=TRUE', conditions), 'TRUE', 'FALSE'),
         'Normalization' = ifelse(grepl('phosphoscore_corr_', conditions), 'TRUE', 'FALSE'),
         'Residues on activity' = ifelse(grepl('act', conditions), 'TRUE', 'FALSE'))

# Define colors for each technical condition
database <- c('Dorothea', 'Dorothea+SIGNOR', 'CollecTRI', 'CollecTRI+SIGNOR')
paired_pal <- RColorBrewer::brewer.pal(n = 12, 'Paired')
database_colors <- paired_pal[7:10]
names(database_colors) <- database

Rpal <- RColorBrewer::brewer.pal(n = 8, 'Dark2')

my_colour = list(
  'Database' = database_colors,
  'Hypergeometric test' = c('TRUE' = Rpal[1], 'FALSE' = 'lightgrey'),
  'VIPER correction with proteomics' = c('TRUE' = Rpal[4], 'FALSE' = 'lightgrey'),
  'Normalization' = c('TRUE' = Rpal[5], 'FALSE' = 'lightgrey'),
  'Residues on activity' = c('TRUE' = Rpal[8], 'FALSE' = 'lightgrey')
)

# ==== Draw gold standard heatmap  ==== #
result <- draw_benchmark_heatmap(NES_df = NES_all,
                                 gold_standard = regulators,
                                 backbone_annotation = annotation_df_plot,
                                 annotation_colors = my_colour)
```

### Kinases and Phosphatases

```{r  fig.height = 4, fig.width = 8, fig.align='center'}
# Read the kinases and phosphatases inferred with KSEA, PhosphoScore or both
# and transform the tibble in a matrix without NAs
all_com_df <- read_tsv(paste0(output_dir, 'kinphos_protein_total.tsv'))
NES_all <- all_com_df %>% dplyr::select(gene_name, contains('NES'))
NES_mat <- NES_all %>% column_to_rownames('gene_name')
NES_mat[is.na(NES_mat)] <- 0

# Define annotation tables for the different technical conditions
annotation_df <- tibble(conditions = colnames(NES_all)[-1], sample = paste0('Sample', 1:(ncol(NES_all)-1)))

annotation_df_plot <- annotation_df %>%
  mutate('Normalization' = ifelse(grepl('NES_corr_', conditions), 'TRUE', 'FALSE'),
         'Database' = ifelse(grepl('_no_int_', conditions), 'Omnipath', 'Omnipath+Kinome Atlas'),
         'Hypergeometric test' = ifelse(grepl('weight=TRUE', conditions), 'TRUE', 'FALSE'),
         'VIPER correction with proteomics' = ifelse(grepl('prot=TRUE', conditions), 'TRUE', 'FALSE'),
         'Residues on activity' = ifelse(grepl('act', conditions), 'TRUE', 'FALSE'))

# Define colors for each technical condition
database <- c('Omnipath', 'Omnipath+Kinome Atlas')
paired_pal <- RColorBrewer::brewer.pal(n = 12, 'Paired')
database_colors <- paired_pal[5:6]
names(database_colors) <- database

Rpal <- RColorBrewer::brewer.pal(n = 8, 'Dark2')

my_colour = list(
  'Database' = database_colors,
  'Hypergeometric test' = c('TRUE' = Rpal[1], 'FALSE' = 'lightgrey'),
  'VIPER correction with proteomics' = c('TRUE' = Rpal[4], 'FALSE' = 'lightgrey'),
  'Normalization' = c('TRUE' = Rpal[5], 'FALSE' = 'lightgrey'),
  'Residues on activity' = c('TRUE' = Rpal[8], 'FALSE' = 'lightgrey')
)

result <- draw_benchmark_heatmap(NES_df = NES_all,
                                 gold_standard = regulators,
                                 backbone_annotation = annotation_df_plot,
                                 annotation_colors = my_colour)
```

### Other signaling proteins

```{r  fig.height = 4, fig.width = 8, fig.align='center'}

# Read the PhosphoScore table without kinases and transcription factors 
# and transform the tibble in a matrix without NAs

# Read whole final_score table
NES_all <- read_tsv(paste0(output_dir, 'other_protein_total.tsv'))
NES_mat <- NES_all %>% column_to_rownames('gene_name')
myBreaks <- c(seq(min(NES_mat, na.rm = TRUE), 0, length.out=ceiling(paletteLength/2) +1),
              seq(max(NES_mat, na.rm = TRUE)/paletteLength, max(NES_mat, na.rm = TRUE), length.out=floor(paletteLength/2)))
NES_mat[is.na(NES_mat)] <- 0

# Define annotation tables for the different technical conditions
annotation_df <- tibble(conditions = colnames(NES_all)[-1], sample = paste0('Sample', 1:(ncol(NES_all)-1)))

annotation_df_plot <- annotation_df %>%
  mutate('Normalization' = ifelse(grepl('phosphoscore_corr_', conditions), 'TRUE', 'FALSE'),
         'Residues on activity' = ifelse(grepl('act', conditions), 'TRUE', 'FALSE'))

# Define colors for each technical condition
Rpal <- RColorBrewer::brewer.pal(n = 8, 'Dark2')

my_colour = list(
  'Normalization' = c('TRUE' = Rpal[5], 'FALSE' = 'lightgrey'),
  'Residues on activity' = c('TRUE' = Rpal[8], 'FALSE' = 'lightgrey')
)

result <- draw_benchmark_heatmap(NES_df = NES_all,
                                 gold_standard = regulators,
                                 backbone_annotation = annotation_df_plot,
                                 annotation_colors = my_colour)
```

# Visualize the best combination of parameters

Creation of **best result table** of benchmarking of Step1

```{r}
source('0.libraries.R')

# Selection of tables of inferred proteins with best parameters
tf_df <- read_tsv('./results/3.protein_activity_inference/combination/tfs_protein_total.tsv') %>%
  dplyr::select(gene_name,
                final_score = 'NES_DoroSign_weight=TRUE_prot=FALSE_phosphoscore_corr_phospho_act',
                method = method_19) %>%
  mutate(mf = 'tf') %>%
  filter(!is.na(final_score))

kin_phos_df <- read_tsv('./results/3.protein_activity_inference/combination/kinphos_protein_total.tsv') %>%
  dplyr::select(gene_name,
                final_score = 'NES_corr_phospho_int_weight=TRUE_prot=TRUE_phosphoscore_corr_phospho_act',
                method = method_1) %>% filter(!is.na(final_score))

kin_phos_df <- SignalingProfiler::molecular_function_annotation(kin_phos_df, 'human')

other_df <- read_tsv('./results/3.protein_activity_inference/combination/other_protein_total.tsv') %>%
  dplyr::select(gene_name, final_score = phosphoscore_corr_phospho_act) %>%
  filter(!is.na(final_score)) %>%
  mutate(mf = 'other',
         method = 'PhosphoScore')


# create a final inferred activity dataframe
inferred_proteins_df <- bind_rows(tf_df, kin_phos_df, other_df)

inferred_proteins_df

write_tsv(inferred_proteins_df,
          './results/3.protein_activity_inference/best_inferred_proteins.tsv')
```

### Visualization of the best result of the benchmarking of Step1

```{r fig.align='center', fig.width=5, fig.height=7}
inferred_proteins_df <- read_tsv('./results/3.protein_activity_inference/best_inferred_proteins.tsv')

inferred_proteins_df$mf <- factor(inferred_proteins_df$mf, levels = c('tf', 'kin', 'phos', 'other'))

inferred_proteins_df %>% distinct() %>% count(mf)

ggplot(inferred_proteins_df,
       aes(x = fct_reorder(gene_name, final_score), y = final_score)) +
  geom_bar(stat = "identity", aes(fill = final_score), alpha = 1) +
  facet_grid(mf ~ ., space ='free_y', scales = 'free') +
  theme_classic() +
  scale_fill_gradient2(low = "#89ABD6", high = "#FF6A72",
                       mid = "whitesmoke", midpoint = 0) +
  theme(text = element_text(size= 6),
        axis.text.x = element_text(angle = 0, vjust = 1, size = 5),
        legend.position="bottom",
        legend.title = element_blank()) +
  xlab('') +
  coord_flip()+
  ylab('Final score') 

```

**Figure 3C** Bar plot displaying the activity modulation (metformin-treated vs control condition) for transcription factors, kinases, phosphatases, and other signaling proteins in the top result from Step 1. Blue and red represent inactive and active protein, respectively.


```{r fig.align='center', fig.width=3, fig.height=3}
# Read best inferred proteins
inferred_proteins_df <- read_tsv('./results/3.protein_activity_inference/best_inferred_proteins.tsv')

methods_color <-(c( "#99B9D8", "#A6A8CD", "#F4B47F"))
names(methods_color) <- c('VIPER', 'VIPER;PhosphoScore', 'PhosphoScore')

ggplot(inferred_proteins_df, aes(x = mf, fill = method)) +
  geom_bar(stat = "count") +
  geom_text(stat = 'count', aes(label = after_stat(count)),
            position = position_stack(vjust = 1.05), size = 3, color = 'black') +
  #scale_fill_brewer(palette = 'Set2') +
  scale_fill_manual(values = methods_color) +
  theme_classic() +
  theme(text = element_text(size=10),
        axis.text.x = element_text(angle = 0, vjust = 1, size = 8),
        legend.position="bottom",
        legend.title = element_blank()) +
  xlab('') +
  ylab('N° of inferred proteins')


```

**Figure 3D** Bar plot representing the proportion of transcription factors, kinases, phosphatases and other signaling proteins identified using PhosphoScore (orange) or footprint-based methods (blue).
